aten/src/ATen/native/quantized/cpu/qpool.cpp

```diff
@@ -0,0 +1,205 @@
+          while(w_start < 0) w_start += dW;
+
+          // local pointers
+          T* o_p = oData + p * oW * oH + row * oW + col;
+          int64_t* ind_p = index + p * oW * oH + row * oW + col;
+
+          // local max
+          int64_t max_index = -1;
+          auto max_val = std::numeric_limits<typename T::underlying>::lowest();
+          int64_t tcntr = 0;  // center point
+          int64_t x, y;
+          for (y = h_start; y < h_end; y += dH) {
+            for (x = w_start; x < w_end; x += dW) {
+              tcntr = y * iW + x;
+              auto val = (i_p + tcntr)->val_;
+              if ((val > max_val) || std::isnan(val)) {
+                max_val = val;
+                max_index = tcntr;
+              }
+            }
+          }
+          // set output
+          *o_p = T(max_val);
+          // store for backprop.
+          *ind_p = max_index;
+        }
+      }
+    }
+  });
+}
+
+template <typename Q>
+Tensor q_maxpool_2d(Tensor qx,                  // Input Tensor (Quantized)
+                    int64_t kH, int64_t kW,     // kernel size
+                    int64_t sH, int64_t sW,     // strides
+                    int64_t dH, int64_t dW,     // dilation
+                    int64_t pH, int64_t pW) {   // padding
+  // Check input dimensions.
+  TORCH_CHECK(kH > 0 && kW > 0, "kernel_size should be greater than zero.");
+  TORCH_CHECK(sH > 0 && sW > 0, "strides should be greater than zero.");
+  TORCH_CHECK(dH > 0 && dW > 0, "dilation should be greater than zero.");
+
+  int ndim = qx.dim();
+  TORCH_CHECK(ndim == 3 || ndim == 4,
+             "Expecting the input tensor of rank 3 or 4.");
+  int dimc = 0;
+  int dimh = 1;
+  int dimw = 2;
+  int nbatch = 1;
+  if (ndim == 4) {  // Includes batches
+    ++dimc;
+    ++dimh;
+    ++dimw;
+    nbatch = qx.size(0);
+  }
+
+  // Check if inputs are valid.
+  int64_t iC = qx.size(dimc);
+  int64_t iH = qx.size(dimh);
+  int64_t iW = qx.size(dimw);
+  TORCH_CHECK(iC > 0 && iH > 0 && iW > 0,
+             "input dimensions must be non-zero.");
+  TORCH_CHECK(qx.numel() > 0 && (ndim == 3 || ndim == 4),
+             "non-empty 3D or 4D input tensor is expected.");
+  TORCH_CHECK(kH/2 >= pH && kW/2 >= pW,
+             "padding should be smaller than half of kernel_size.");
+
+  // Check output dimensions.
+  int64_t oC = iC;
+  int64_t oH = pooling_output_shape(iH, kH, pH, sH, dH, false);
+  int64_t oW = pooling_output_shape(iW, kW, pW, sW, dW, false);
+  TORCH_CHECK(oH > 0 && oW > 0, "the resulting Tensor is too small.");
+
+  std::vector<int64_t> oSizes;
+  if (ndim == 3) {
+    oSizes = {oC, oH, oW};
+  } else {
+    oSizes = {nbatch, oC, oH, oW};
+  }
+
+  Tensor qy = at::_empty_affine_quantized(
+    oSizes,
+    qx.options().dtype(toQIntType(qx.scalar_type())),
+    qx.q_scale().toDouble(),
+    qx.q_zero_point().toLong());
+  Q* qxd;
+  if (qx.is_contiguous()) {
+    qxd = qx.data<Q>();
+  } else {
+    qxd = qx.contiguous().data<Q>();
```

Comment:
this is a bug. `.contiguous()` returns a new tensor that immediately gets destroyed but you keep its memory pointer. Also it'd return you the same instance if it's already contigutous.

So just:
```
qxc = qx.contiguous();
Q* qxd = qxc.data<Q>();
```